### DISCUSSION TITLE
How do I define an constant variable with EsBuild?

### DISCUSSION THREAD

**Discussion by Pentadome** [vite-pwa/vite-plugin-pwa#762] [Q&A] [answered]
How do I define an constant variable with EsBuild?

I tried the following:

```ts
export const pwaConfig: Partial<VitePWAOptions> = {
  registerType: 'autoUpdate',
  devOptions: {
    enabled: true,
    type: 'module',
  },
  strategies: 'injectManifest',
  injectManifest: {
    swSrc: 'pwa/customServiceWorker.ts',
    // precache these files.
    globPatterns: ['**/*.{js,css,html,svg,woff,woff2,png}'],
  },
  integration: {
    configureCustomSWViteBuild: (options) => {
      throw 'test';  // <- never reached!
      if (!options.esbuild) {
        options.esbuild = {};
      }
      options.esbuild.define ??= {};
      options.esbuild.define.__FIREBASE_CONFIG = `'${process.env.PUBLIC_FIREBASE_CONFIG}'`;
    },
  },
}
```

But it seems I am not understanding what `configureCustomSWViteBuild` is for. It seems to never get called. How do I define an esbuild variable?

**Comment by userquin** [comment]
Are you using any integration (Nuxt, SvelteKit, Astro...)?

**Reply by Pentadome** [reply]
Yea i am using Nuxt

**Reply by userquin** [reply]
Have you tried running Nuxt build/generate script? In dev server there is no build (when using `injectManfest` strategy), Vite will serve your custom service worker via esbuild.

You can use Vite `define` or `.env` files accessing them via`import.meta.env` from your service worker: https://vite-pwa-org.netlify.app/guide/change-log.html#new-vite-build

Rigth now, when using PWA Assets the PWA integration will be override by the Nuxt PWA plugin.

**Reply by userquin** [reply]
https://github.com/vite-pwa/vite-plugin-pwa/blob/main/src/vite-build.ts#L90

**Reply by Pentadome** [reply]
```ts
const pwaConfig: Partial<VitePWAOptions> = {
  registerType: 'autoUpdate',
  devOptions: {
    enabled: true,
    type: 'module',
  },
  strategies: 'injectManifest',
  srcDir: 'pwa',
  filename: 'customServiceWorker.ts',
  injectManifest: {
    swSrc: 'pwa/customServiceWorker.ts',
    // precache these files.
    globPatterns: ['**/*.{js,css,html,svg,woff,woff2,png}'],
  },
  integration: {
    configureCustomSWViteBuild: (options) => {
      options.define ??= {};
      options.define.__FIREBASE_CONFIG = `'${process.env.PUBLIC_FIREBASE_CONFIG}'`;
    },
  },
}
```
I tried this, but unfortunately  during development i get `Uncaught SyntaxError: Cannot use 'import.meta' outside a module` and in production import.meta.__FIREBASE_CONFIG is replaced with undefined. But when I console.log process.env.PUBLIC_FIREBASE_CONFIG, it is there.

**Reply by userquin** [reply]
can you provide a minimal repoduction or a link to your repo if public?

**Reply by userquin** [reply]
can you check if you've `"type": "module"` entry in your package.json?

**Reply by Pentadome** [reply]
>can you provide a minimal repoduction or a link to your repo if public?

sure

https://stackblitz.com/edit/github-4afvcv?file=package.json

I don't think you can test service workers on stackblitz tho. But you can download it from there.

> can you check if you've "type": "module" entry in your package.json?
yea i got type "module"

**Reply by userquin** [reply, marked as answer]
You should:
- add `/// <reference types="vite/client" />` to your service worker
- add `.env` file and prefix public entries with `VITE_` and use `import.meta.env.VITE_XXX`: there is no way to access any `process.env` variable from the service worker (Nuxt will not expose any `vite.define` nor `runtimeConfig.public` in `import.meta.env` (also checked `vite.define.envPrefix = ['VITE_', 'NUXT_']`)

https://stackblitz.com/edit/github-4afvcv-xu75vp?file=nuxt.config.ts,pwa%2FcustomServiceWorker.ts,.env

![imagen](https://github.com/user-attachments/assets/28776022-7f69-47e0-b320-dd855200233d)

**Reply by userquin** [reply]
Beware when deploying to your server, check it is working correctly, you should check any private stuff not being exposed (Vite should filter out any entry not starting with VITE_)

**Reply by userquin** [reply]
You can check https://vite.dev/guide/env-and-mode to add proper `.env` files for your local, pre and prod environments.

**Reply by Pentadome** [reply]
Thanks a lot. I got it working now 👍

**Reply by userquin** [reply]
@Pentadome Daniel Roe response:
> nuxt prefers runtime config. we don’t expose process.env
> we run in non node environments so can’t rely on process
